src/test/java/com/cloudbees/jenkins/plugins/sshagent/SSHAgentStepWorkflowTest.java

```diff
@@ -0,0 +1,144 @@
+                credentialIds.add(CREDENTIAL_ID);
+
+                SSHUserPrivateKey key = new BasicSSHUserPrivateKey(CredentialsScope.GLOBAL, credentialIds.get(0), "cloudbees",
+                        new BasicSSHUserPrivateKey.DirectEntryPrivateKeySource(getPrivateKey()), "cloudbees", "test");
+                SystemCredentialsProvider.getInstance().getCredentials().add(key);
+                SystemCredentialsProvider.getInstance().save();
+
+                WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "sshAgentAvailableAfterRestart");
+                p.setDefinition(new CpsFlowDefinition(""
+                        + "node {\n"
+                        + "  sshagent (credentials: ['" + CREDENTIAL_ID + "']) {\n"
+                        + "    sh 'ssh -o StrictHostKeyChecking=no -p " + getAssignedPort() + " -v -l cloudbees " + SSH_SERVER_HOST + "'\n"
+                        + "    echo \"SSH Agent before restart ${env.SSH_AUTH_SOCK}\"\n"
+                        + "    semaphore 'sshAgentAvailableAfterRestart'\n"
+                        + "    sh 'ssh -o StrictHostKeyChecking=no -p " + getAssignedPort() + " -v -l cloudbees " + SSH_SERVER_HOST + "'\n"
+                        + "    echo \"SSH Agent after restart ${env.SSH_AUTH_SOCK}\"\n"
+                        + "  }\n"
+                        + "}\n", true));
+                // get the build going
+                WorkflowRun b = p.scheduleBuild2(0).getStartCondition().get();
+                CpsFlowExecution e = (CpsFlowExecution) b.getExecutionPromise().get();
+
+                // wait until the executor gets assigned and the execution pauses
+                SemaphoreStep.waitForStart("sshAgentAvailableAfterRestart/1", b);
+                assertTrue(JenkinsRule.getLog(b), b.isBuilding());
+            }
+        });
+        story.addStep(new Statement() {
+            @Override
+            public void evaluate() throws Throwable {
+                WorkflowJob p = story.j.jenkins.getItemByFullName("sshAgentAvailableAfterRestart", WorkflowJob.class);
+                WorkflowRun b = p.getBuildByNumber(1);
+
+                SemaphoreStep.success("sshAgentAvailableAfterRestart/1", null);
+
+                story.j.assertBuildStatusSuccess(story.j.waitForCompletion(b));
+
+                Pattern pattern = Pattern.compile("jenkins([0-9])+\\.jnr");
+                Scanner sc = new Scanner(b.getLogFile());
+                List<String> socketFile = new ArrayList<String>();
+                while (sc.hasNextLine()) {
+                    String match = sc.findInLine(pattern);
+                    if (match != null) {
+                        socketFile.add(match);
+                    } else {
+                        sc.nextLine();
+                    }
+                }
+                sc.close();
+
+                assertEquals(socketFile.size(), 2);
+                assertNotEquals(socketFile.get(0), socketFile.get(1));
```

Comment:
Works, though this is testing the design of the current implementation—that one `$SSH_AUTH_SOCK` is set initially, then a different one after the restart—rather than the user feature—that both `ssh` commands work, even if the agent is killed during the restart. A subtle difference, but consider what would happen if you improved the implementation in the future to use a single socket location, which would fix the

```groovy
node {
  sshagent(…) {
    sh '''
sleep 999 # ← restart Jenkins here
ssh …
'''
  }
}
```

bug. After that improvement, this test would fail, even though the behavior was actually better. So a better test would ignore the implementation details about this environment variable, instead just running the `ssh` command before and after the restart, and being sure to kill the server before the restart; this would fail as expected if you forgot to implement `onResume`.

An even better implementation would of course handle this bug

```groovy
node {
  sshagent(…) {
    sh '''
sleep 999
# ← restart Jenkins here
ssh …
'''
  }
}
```

by keeping an agent running continuously across the restart, by forking an external process. You would have to edit the test a little to match such an implementation, of course.